### PR TITLE
Fix dropout when using a new-style rng

### LINKED ISSUE #244: [Issue]: TE dropout requires old-style PRNGs
### Problem Description

Jax changed the preferred type for RNG seeds from an array of ints to an opaque type in [August 2023](https://docs.jax.dev/en/latest/jep/9263-typed-keys.html#overview).

TE attention still requires the old-style keys in [transformer_engine/jax/cpp_extensions/attention.py](https://github.com/ROCm/TransformerEngine/blob/c97ecae53e8cd1d8b925530eef25036ef43f9181/transformer_engine/jax/cpp_extensions/attention.py#L208).
Any applications using new-style keys will see an error when this line of code attempts to convert the seed to an int.

It may be possible to work around this issue easily by changing the line to:
`seed = jax.random.key_data(seed).astype(self.rng_state_dtype)`

### Operating System

Ubuntu 22.04, Ubuntu 24.04

### CPU

Any

### GPU

MI300x series

### ROCm Version

Any 6.4

### ROCm Component

_No response_

### Steps to Reproduce

Run TE DotProductAttention with dropout != 0.0 (for example set it to 0.2) and a PRNG created with  jax.random.key() instead of jax.random.PRNGKey()

### (Optional for Linux users) Output of /opt/rocm/bin/rocminfo --support

_No response_

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by gabeweisz**
# Description

Previously TE would crash with a type error [at this line of code](https://github.com/ROCm/TransformerEngine/blob/c97ecae53e8cd1d8b925530eef25036ef43f9181/transformer_engine/jax/cpp_extensions/attention.py#L208) if new-style RNGs were used. This PR fixes the error by converting a new-style RNG to an int properly.

Fixes #244 

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

## Changes

Please list the changes introduced in this PR:

- Uses Jax's [recommended mechanism](https://docs.jax.dev/en/latest/jep/9263-typed-keys.html) to convert a new-style RNG to int data when a new-style RNG is used
- Includes tests for new-style RNGs

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by gabeweisz** [comment]
@wangye805 - this RNG-fixing PR is ready. Do you have any comments on it?

File: tests/jax/test_fused_attn.py [hunk 1]
```diff
@@ -931,6 +936,13 @@ def check_dqkv(primitive, reference, pad, idx):
         pytest.param(0.1, id="DROP_0.1"),
     ],
 )
+@pytest.mark.parametrize(
+    "use_old_rng",
+    [
+        pytest.param(True, id="Old-style rng"),
+        pytest.param(False, id="New-style rng"),
+    ],
```
Comment (Micky774) [inline tests/jax/test_fused_attn.py:954]: Adding this as an additional parameter doubles the test parameter space. Is this strictly necessary, or might it be sufficient to add a singular non-regression test to check the exact error that this PR aims to fix?

File: transformer_engine/jax/cpp_extensions/attention.py [hunk 2]
```diff
@@ -205,6 +205,10 @@ def check_seed(self, seed, dropout_probability, is_training):
                 f"casted to dtype {self.rng_state_dtype}. "
                 "Please use threefry/rbg/unsafe_rbg PRNG implementations to remove this warning."
             )
+            if jnp.issubdtype(seed.dtype, jax.dtypes.prng_key):
+                # New-style RNGs cannot directly be cast to integers
+                # See https://docs.jax.dev/en/latest/jep/9263-typed-keys.html for details
+                seed = jax.random.key_data(seed)
```
Comment (Micky774) [inline transformer_engine/jax/cpp_extensions/attention.py]: While this serves as an immediate hotfix, we probably should address the underlying problem and update to avoid relying on directly accessing the raw buffer. Is this necessary due to a technical restraint, or is it just going to require some effort to update our consumption of the random seed?

File: tests/jax/test_fused_attn.py (hunk 1 above)
Comment (gabeweisz) [reply tests/jax/test_fused_attn.py:954]: I thought it would be useful to test everything, but I can reduce the number of tests that use this. Since the error is a type-cast error, we probably only need one

File: transformer_engine/jax/cpp_extensions/attention.py (hunk 2 above)
Comment (gabeweisz) [reply transformer_engine/jax/cpp_extensions/attention.py]: My guess is that it would require a significant effort to change CK / AITER C++ libraries to change the consumption of the random seed.

Comment (gabeweisz) [reply transformer_engine/jax/cpp_extensions/attention.py]: CK's C++ code requires a [64-bit int as a random seed](https://github.com/ROCm/composable_kernel/blob/b4f3487d8423a0e14bbb90e1cf8305d3560f3d17/example/ck_tile/01_fmha/fmha_fwd.cpp#L424). I am sure AITER is similar. I do not think Jax opaque PRNGs can be passed to C++ as a 64-bit int, but if you think they can we can report this to the CK/AITER team

Comment (gabeweisz) [reply transformer_engine/jax/cpp_extensions/attention.py]: We could potentially use randint() instead of jax.random.key_data() to get seed for CK. I am not sure which version is better, but this version is suggested in the Jax docs

File: tests/jax/test_fused_attn.py [hunk 3]
```diff
@@ -387,7 +388,10 @@ def _setup_inputs(self):
         self.cp_size = self.mesh.shape.get(self.mesh_resource.cp_resource, 1)
         self.tp_size = self.mesh.shape.get(self.mesh_resource.tp_resource, 1)
 
-        key = jax.random.PRNGKey(0)
+        if self.use_old_rng:
```
Comment (wangye805) [inline tests/jax/test_fused_attn.py]: Can we use the is_hip_extension() to guard our rocm side change?

File: tests/jax/test_fused_attn.py [hunk 4]
```diff
@@ -583,7 +587,7 @@ def generate_random_segment_ids(
                 case _:
                     raise ValueError(f"Unknown {self.seq_desc_format=}")
 
-        self.dropout_rng = dropout_key if self.dropout_prob > 0 else None
+        self.dropout_rng = dropout_key
```
Comment (wangye805) [inline tests/jax/test_fused_attn.py]: Why do we need dropout_key when dropout_prob ==0 ?

File: tests/jax/test_fused_attn.py [hunk 5]
```diff
@@ -626,9 +630,10 @@ def to_dp_shardings(x):
             self.bias_pspec = PartitionSpec()
         self.bias_sharding = NamedSharding(self.mesh, self.bias_pspec)
 
-        self.dropout_rng_pspec = PartitionSpec(
-            None,
-        )
+        if jnp.issubdtype(self.dropout_rng.dtype, jax.dtypes.prng_key):
```
Comment (wangye805) [inline tests/jax/test_fused_attn.py]: Same here, please use the is_hip_extension to guard the rocm side changes

File: tests/jax/test_fused_attn.py [hunk 6]
```diff
@@ -858,7 +863,7 @@ def check_dqkv(primitive, reference, pad, idx):
 
             # Assume all batch has the same actual_seqlen, probably needs to extend the tests
             bias_mask = self.mask[0, 0]
-            
+
```
Comment (wangye805) [inline tests/jax/test_fused_attn.py]: Let's not introduce unnecessary reformatting

File: transformer_engine/jax/cpp_extensions/attention.py [hunk 7]
```diff
@@ -205,6 +205,10 @@ def check_seed(self, seed, dropout_probability, is_training):
                 f"casted to dtype {self.rng_state_dtype}. "
                 "Please use threefry/rbg/unsafe_rbg PRNG implementations to remove this warning."
             )
+            if jnp.issubdtype(seed.dtype, jax.dtypes.prng_key):
```
Comment (wangye805) [inline transformer_engine/jax/cpp_extensions/attention.py]: Same here, use is_hip_extension to guard rocm side changes

**Review by wangye805 — CHANGES_REQUESTED** [review]

File: tests/jax/test_fused_attn.py (hunk 3 above)
Comment (gabeweisz) [reply tests/jax/test_fused_attn.py]: I will make that change - but I will point out that this code is broken on Nvidia TE too

File: tests/jax/test_fused_attn.py (hunk 4 above)
Comment (gabeweisz) [reply tests/jax/test_fused_attn.py:590]: There was some code that failed when dropout_key was none

File: tests/jax/test_fused_attn.py (hunk 6 above)
Comment (gabeweisz) [reply tests/jax/test_fused_attn.py]: My text editor did this - this was not on purpose

File: tests/jax/test_fused_attn.py (hunk 4 above)
Comment (gabeweisz) [reply tests/jax/test_fused_attn.py]: The check on line 633 failed when it was non - I'll change it there

File: tests/jax/test_fused_attn.py (hunk 3 above)
Comment (Micky774) [reply tests/jax/test_fused_attn.py]: That makes sense. @wangye805 do you think we can do without the hip guard in that case? Especially seeing as it would break on CPU backend too?

**Review by wangye805 — APPROVED** [review]
